From e9771d7eb526a7c24512aaaee48cc74a357c5082 Mon Sep 17 00:00:00 2001
From: Phil Steitz <psteitz@apache.org>
Date: Mon, 21 Sep 2009 01:30:53 +0000
Subject: [PATCH] Fixed implementation of
 EmpiricalDistributionImpl#getUpperBounds to match interface contract.  Added
 getGeneratorUpperBounds method to EmpiricalDistributionImpl providing
 previous behavior.

Jira: Math-298

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/math/trunk@817128 13f79535-47bb-0310-9956-ffa450edef68
---
 .../math/random/EmpiricalDistribution.java       |  2 +-
 src/site/xdoc/changes.xml                        |  5 +++++
 .../math/random/EmpiricalDistributionTest.java   | 16 +++++++++++++++-
 3 files changed, 21 insertions(+), 2 deletions(-)

diff --git a/src/main/java/org/apache/commons/math/random/EmpiricalDistribution.java b/src/main/java/org/apache/commons/math/random/EmpiricalDistribution.java
index 5d2c33a92e..30abf121bc 100644
--- a/src/main/java/org/apache/commons/math/random/EmpiricalDistribution.java
+++ b/src/main/java/org/apache/commons/math/random/EmpiricalDistribution.java
@@ -124,7 +124,7 @@ public interface EmpiricalDistribution {
     /**
      * Returns the array of upper bounds for the bins.  Bins are: <br/>
      * [min,upperBounds[0]],(upperBounds[0],upperBounds[1]],...,
-     *  (upperBounds[binCount-1],max].
+     *  (upperBounds[binCount-2], upperBounds[binCount-1] = max].
      *
      * @return array of bin upper bounds
      */
diff --git a/src/site/xdoc/changes.xml b/src/site/xdoc/changes.xml
index 19d8e6b4fa..150fb6eca0 100644
--- a/src/site/xdoc/changes.xml
+++ b/src/site/xdoc/changes.xml
@@ -39,6 +39,11 @@ The <action> type attribute can be add,update,fix,remove.
   </properties>
   <body>
     <release version="2.1" date="TBD" description="TBD">
+      <action dev="psteitz" tyoe="fix" issue="MATH-298">
+        Fixed implementation of EmpiricalDistributionImpl#getUpperBounds to match
+        interface contract.  Added getGeneratorUpperBounds method to
+        EmpiricalDistributionImpl providing previous behavior.
+      </action>
       <action dev="luc" type="fix" issue="MATH-293" due-to="Benjamin McCann">
         Fixed a OutOfBoundException in simplex solver when some constraints are tight.
       </action>
diff --git a/src/test/java/org/apache/commons/math/random/EmpiricalDistributionTest.java b/src/test/java/org/apache/commons/math/random/EmpiricalDistributionTest.java
index f1b4e07457..850caf7515 100644
--- a/src/test/java/org/apache/commons/math/random/EmpiricalDistributionTest.java
+++ b/src/test/java/org/apache/commons/math/random/EmpiricalDistributionTest.java
@@ -118,7 +118,7 @@ public void testDoubleLoad() throws Exception {
           (empiricalDistribution2.getSampleStats().getStandardDeviation(),
                 1.0173699343977738,10E-7);
 
-        double[] bounds = empiricalDistribution2.getUpperBounds();
+        double[] bounds = ((EmpiricalDistributionImpl) empiricalDistribution2).getGeneratorUpperBounds();
         assertEquals(bounds.length, 100);
         assertEquals(bounds[99], 1.0, 10e-12);
 
@@ -226,6 +226,20 @@ public void testLoadNullFile() throws Exception {
             fail("wrong exception caught");
         }
     }
+    
+    /**
+     * MATH-298
+     */
+    public void testGetBinUpperBounds() {
+        double[] testData = {0, 1, 1, 2, 3, 4, 4, 5, 6, 7, 8, 9, 10};
+        EmpiricalDistributionImpl dist = new EmpiricalDistributionImpl(5);
+        dist.load(testData);
+        double[] expectedBinUpperBounds = {2, 4, 6, 8, 10};
+        double[] expectedGeneratorUpperBounds = {4d/13d, 7d/13d, 9d/13d, 11d/13d, 1};
+        double tol = 10E-12;
+        TestUtils.assertEquals(expectedBinUpperBounds, dist.getUpperBounds(), tol);
+        TestUtils.assertEquals(expectedGeneratorUpperBounds, dist.getGeneratorUpperBounds(), tol);  
+    }
 
     private void verifySame(EmpiricalDistribution d1, EmpiricalDistribution d2) {
         assertEquals(d1.isLoaded(), d2.isLoaded());
